### ISSUE TITLE
Add Host/Host Templates Improvements

### ISSUE THREAD

**Issue by cjellick** [rancher/rancher#8703]
#### Functional
- Users should be able to create hosts in rancher by simply selecting a flavor like "Digitial Ocean with 2 gigs of memory"
- If a machine driver config property cannot be wrapped into the flavor (ie, access and secret keys), the user shoudl be able to securely save those credentials in Rancher for reuse when creating machines.

#### Technical
Ensure flavor-machine-driver is fully functioning and free of defects
- The driver is already implemented, but needs to be more fully tested and if bugs come up, they need addressed
- We need a mechanism for pushing the provider and flavor config files down to GMS
- We need to see if there are any other underlying cloud provider drivers that need implemented. Currently have packet, aws, and digital ocean

Ensure go-machine-services integration with the cattle's hostTemplates is fully functioning and defect free
- Users can either explicitly pick a hostTemplate to use when creating a machine or if they don't, gms should pick one based off the name of flavor being used. HostTemplate has a field called `flavorPrefix`. The idea is that if the user has not selected a host template, you look at the prefix of the flavor which will be named like `digitalocean-2gb-sfo1`. In this case, the flavor prefix is digitalocean, so you would find a HostTemplate in the Rancher API whose `flavorPrefix` field has a value of "digitalocean".
- So, there are two use cases: 1) user explicitly picks a hostTemplate 2) gms infers a hostTemplate. Use case 1 is implemented but not tested at all, so it must be tested and verified by developer before handing off to QA. Use case 2 is not implemented at all, so it needs implemented.

### Random
- HostTemplate is a resource in the v2-beta API. You can hit it using the API UI.
- The new UI @vincent99 is working on has already integrated with hostTemplate and the flavor driver. You can switch to that to see how it works

**Comment by imikushin** [comment]
Got some questions:
- a mechanism for pushing the provider and flavor config files down to GMS:
  - who will be adding/removing flavors, at what time? (e.g. admin - at run time, or us - as part of rancher distribution)
  - will there be a catalog of flavors?
- gms infers a hostTemplate:
  - what do we do with the hostTemplate(s) when we've found one (or more)? can I extract the account credentials from a hostTemplate?
- credential store:
  - does rancher have it or is building one a part of this?

I think these should go into separate (related) issues:
- implement the initial set of flavors (for digitalocean, aws, packet)
- gce driver and flavors

**Comment by cjellick** [comment]
> - a mechanism for pushing the provider and flavor config files down to GMS:
>   - who will be adding/removing flavors, at what time? (e.g. admin - at run time, or us - as part of rancher distribution)
>   - will there be a catalog of flavors?

I learned from @cloudnautique  today that, at least for the alpha, the flavors will be managed by him as part of the deployment process. So, you have no work to do here. The "catalog" of flavors will be whatever we come up with as wanting to ship.

> - gms infers a hostTemplate:
>   - what do we do with the hostTemplate(s) when we've found one (or more)? can I extract the account credentials from a hostTemplate?

Once you find the template, you apply it to the machine, exactly as if it were passed via hostTemplateId. The logic for applying it when set as hostTemplateId is already implemented, so really you just need to add a bit of code that guesses/infers the template if hostTemplateId == nil and then let the current apply logic take place
You dont need to explicitly extract credentials or any other property, just need to generically apply

> - credential store:
>   - does rancher have it or is building one a part of this?

Credential store is already done and transparent to you. You  just need to find and apply the template and can safely assume we are securely storing the creds

> I think these should go into separate (related) issues:
> - implement the initial set of flavors (for digitalocean, aws, packet)
> - gce driver and flavors

That's fine. Let's go with only the existing providers (do, aws, packet) are in scope

**Comment by cjellick** [comment]
Relaeted https://github.com/rancher/rancher/issues/8592 should be fixed as part of this.

**Comment by imikushin** [comment]
I imagine the host creation flow should look like this:

1. The user is presented the list of available flavors and picks one.
2. The user chooses if they want to use the saved credentials for the flavor if entered before (which is the default), or enter a new set of credentials.
3. The host is created.

Correct?

**Comment by cjellick** [comment]
If more than one template exists with a match flavorPrefix to the flavor being used and the user did not explicitly specif hostTemplateId, should gms pick the first or error out.

**Comment by imikushin** [comment]
To summarize what should be done:

- [ ] create host using the flavor driver using credentials from the specified hostTemplateId
- [ ] initial set of providers (digitalocean, aws, packet) and flavors (2 per provider)
- [ ] end-to-end testing

**Comment by imikushin** [comment]
I've just hit this bug with the flavor-driver:

The machine has just been successfully created: 
```
docker-machine create -d rancher --rancher-digitalocean-access-token=XXXXX --rancher-flavor=digitalocean-sfo2-2gb ivan-test-do-flavor1
```

But `docker-machine config` exits with an error:
```
$ docker-machine config ivan-test-do-flavor1
json: cannot unmarshal object into Go struct field Driver.Driver of type drivers.Driver
```

Probably going to take a couple days to troubleshoot.

**Comment by soumyalj** [comment]
@cjellick @imikushin : Are host templates per environment?
Since UI is not available currently, testing will be done by creating host templates manually.

**Comment by cjellick** [comment]
Yes @soumyalj, they are per environment

**Comment by tfiduccia** [comment]
Version - 2.0 master 8/5
Host Templates are implemented in 2.0 builds.  They are on a per cluster (not environment) basis.  Templates are working and will be further tested.  All bugs will go into another issue.  Closing this.
